heat/core/manipulations.py

```diff
@@ -316,6 +329,169 @@ def concatenate(arrays, axis=0):
             return out
 
 
+def diag(a, offset=0):
+    """
+    Extract a diagonal or construct a diagonal array.
+    See the documentation for `heat.diagonal` for more information about extracting the diagonal.
+
+    Parameters
+    ----------
+    a: ht.DNDarray
+        The array holding data for creating a diagonal array or extracting a diagonal.
+        If a is a 1-dimensional array a diagonal 2d-array will be returned.
+        If a is a n-dimensional array with n > 1 the diagonal entries will be returned in an n-1 dimensional array.
+    offset: int, optional
+        The offset from the main diagonal.
+        Offset greater zero means above the main diagonal, below zero is below the main diagonal.
+
+    Returns
+    -------
+    res: ht.DNDarray
+        The extracted diagonal or the constructed diagonal array
+
+    Examples
+    --------
+    >>> import heat as ht
+    >>> a = ht.array([1, 2])
+    >>> ht.diag(a)
+    array([[1, 0],
+           [0, 2]])
+
+    >>> ht.diag(a, offset=1)
+    array([[0, 1, 0],
+           [0, 0, 2],
+           [0, 0, 0]])
+
+    >>> ht.equal(ht.diag(ht.diag(a)), a)
+    True
+    >>> a = ht.array([[1, 2], [3, 4]])
+    >>> ht.diag(a)
+    array([1, 4])
+    """
+    if len(a.shape) > 1:
+        return diagonal(a, offset=offset)
+    elif len(a.shape) < 1:
+        raise ValueError("input array must be of dimension 1 or greater")
+    if not isinstance(offset, int):
+        raise ValueError("offset must be an integer, got", type(offset))
+    if not isinstance(a, dndarray.DNDarray):
+        raise ValueError("a must be a DNDarray, got", type(a))
+
+    # 1-dimensional array, must be extended to a square diagonal matrix
+    gshape = (a.shape[0] + abs(offset),) * 2
+    off, lshape, _ = a.comm.chunk(gshape, a.split)
+
+    # This ensures that the data is on the correct nodes
+    if offset > 0:
+        padding = factories.empty(
+            (offset,), dtype=a.dtype, split=None, device=a.device, comm=a.comm
+        )
+        a = concatenate((a, padding))
+    elif offset < 0:
+        padding = factories.empty(
+            (abs(offset),), dtype=a.dtype, split=None, device=a.device, comm=a.comm
+        )
+        a = concatenate((padding, a))
+    if not a.is_balanced():
```

Comment:
balance_ will call is_balanced before doing anything so this is optional. you can also just call balance_ here